numpy/_core/src/multiarray/unique.cpp

```diff
@@ -0,0 +1,199 @@
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+
+#include <Python.h>
+
+#include <unordered_set>
+#include <functional>
+
+#include <numpy/npy_common.h>
+#include "numpy/arrayobject.h"
+
+// This is to use RAII pattern to handle cpp exceptions while avoiding memory leaks.
+// Adapted from https://stackoverflow.com/a/25510879/2536294
+template <typename F>
+struct FinalAction {
+    FinalAction(F f) : clean_{f} {}
+    ~FinalAction() { clean_(); }
+  private:
+    F clean_;
+};
+
+template <typename F>
+FinalAction<F> finally(F f) {
+    return FinalAction<F>(f);
+}
+
+template<typename T>
+PyObject* unique(PyArrayObject *self)
```

Comment:
NIT:
```suggestion
static PyObject*
unique(PyArrayObject *self)
```
(still trying to use `static` I guess.  We should change the default linking on that, I am not sure we already did...)